File: free.js
expr = {
    type: "add",
    left: {
        type: "value",
        value: 2
    },
    right: {
        type: "mul",
        left: {
            type: "value",
            value: 3
        },
        right: {
            type: "value",
            value: 5
        }
    }
}

// Straight forward implementation
function eval(ast) {
    if (ast.type === "value") {
        return ast.value;
    }

    let left = eval(ast.left);
    let right = eval(ast.right);

    if (ast.type === "mul") {
        return left * right;
    } else if (ast.type === "add") {
        return left + right;
    }
}

console.log("eval:", eval(expr)); // => 17


// FIX
exprFix = {
    type: "fix",
    item: {
        type: "add",
        left: {
            type: "fix",
            item: {
                type: "value",
                value: 2
            }
        },
        right: {
            type: "fix",
            item: {
                type: "mul",
                left: {
                    type: "fix",
                    item: {
                        type: "value",
                        value: 3
                    }
                },
                right: {
                    type: "fix",
                    item: {
                        type: "value",
                        value: 5
                    }
                }
            }
        }
    }
}

function unFix(fix) {
    return fix.item;
}

function mapExpr(fn, ast) {
    if (ast.type !== "value") {
        ast.left = fn(ast.left);
        ast.right = fn(ast.right);
    }

    return ast;  // where recursion stops in foldFix
}

function foldFix(f, fix) {
    const go = (iter) => {
        return f(mapExpr(go, unFix(iter)));
    };
    return go(fix);
}

// AST -> Int
function algebra(ast) {
    switch (ast.type) {
        case "value":
            return ast.value;
        case "add":
            return ast.left + ast.right;
        case "mul":
            return ast.left * ast.right;
    }
}

function evalFix(ast) {
    return foldFix(algebra, ast);
}

console.log("evalFix:", evalFix(exprFix)); // => 17

// FREE
exprFree = {
    type: "free",
    item: {
        type: "add",
        left: {
            type: "pure",
            item: 2
        },
        right: {
            type: "free",
            item: {
                type: "mul",
                left: {
                    type: "pure",
                    item: 3
                },
                right: {
                    type: "pure",
                    item: 5
                }
            }
        }
    }
}


function unFree(free) {
    return free.item;
}

/* copy mapExpr and algebra from fix. js is dynamic so they still work */

function foldFree(phi, free) {
    if (free.type === "pure") {
        return unFree(free);
    }
    // otherwise free.type === "free"
    return phi(mapExpr(x => foldFree(phi, x), unFree(free)));
}

function evalFree(ast) {
    return foldFree(algebra, ast);
}

console.log("evalFree: ", evalFree(exprFree)); // => 17

File: free.hs
-- Straight forward implementation
data AST = Value Int
         | Add AST AST
         | Mul AST AST
         deriving Show


expr :: AST
expr = Add
        (Value 2)
        (Mul
            (Value 3)
            (Value 5))

eval :: AST -> Int
eval (Value n)   = n
eval (Add e1 e2) = eval e1 + eval e2
eval (Mul e1 e2) = eval e1 * eval e2

-- >>> eval expr
-- 17

-- FIX
newtype Fix f = Fix { unFix :: f (Fix f) }

foldFix :: Functor f => (f a -> a) -> Fix f -> a
foldFix f = go where go = f . fmap go . unFix 


data ASTFix r = ValueFix Int
              | AddFix r r
              | MulFix r r
              deriving Show

instance Functor ASTFix where
    fmap _ (ValueFix n) = ValueFix n
    fmap f (AddFix a b) = AddFix (f a) (f b)
    fmap f (MulFix a b) = MulFix (f a) (f b)


exprFix :: Fix ASTFix
exprFix = Fix (AddFix
                (Fix (ValueFix 2))
                (Fix (MulFix 
                        (Fix (ValueFix 3)) 
                        (Fix (ValueFix 5)))))


algebra :: ASTFix Int -> Int
algebra (ValueFix n)   = n
algebra (AddFix a1 a2) = a1 + a2
algebra (MulFix a1 a2) = a1 * a2


evalFix :: Fix ASTFix -> Int
evalFix = foldFix algebra  

-- >>> evalFix exprFix
-- 17

-- Free
data Free f a = Pure a | Free (f (Free f a))

showFree :: (Show a, Functor f, Traversable f) => Free f a -> IO ()
showFree (Pure a)  = print a
showFree (Free fs) = mapM_ showFree fs


instance Functor f => Functor (Free f) where
    fmap f (Pure a)  = Pure (f a)
    fmap f (Free fa) = Free (fmap (fmap f) fa) 


instance Functor f => Applicative (Free f) where
    pure             = Pure
    (Pure f)  <*> fa = fmap f fa
    (Free fs) <*> fa = Free (fmap (<*> fa) fs)


instance Functor f => Monad (Free f) where
    return          = pure 
    (Pure a)  >>= k = k a 
    (Free fs) >>= k = Free (fmap (>>= k) fs)


data ASTFree r = AddFree r r | MulFree r r 

instance Functor ASTFree where
    fmap f (AddFree a b) = AddFree (f a) (f b)
    fmap f (MulFree a b) = MulFree (f a) (f b)


foldFree :: Functor f => (f a -> a) -> Free f a -> a
foldFree _   (Pure a) = a
foldFree phi (Free m) = phi (foldFree phi <$> m)


exprF :: Free ASTFree Int
exprF = Free (AddFree
                (Pure 2)
                (Free (MulFree
                        (Pure 3)
                        (Pure 5))))


algebraF :: ASTFree Int -> Int
algebraF (AddFree a b) = a + b
algebraF (MulFree a b) = a * b


evalFree :: Free ASTFree Int -> Int 
evalFree = foldFree algebraF
